service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectory.java

```diff
@@ -0,0 +1,106 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.hl7.fhir.dstu3.model.Bundle;
+import org.hl7.fhir.dstu3.model.PractitionerRole;
+import org.hl7.fhir.dstu3.model.Reference;
+import org.hl7.fhir.dstu3.model.ResourceType;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import uk.nhs.adaptors.gp2gp.common.service.RandomIdGeneratorService;
+import uk.nhs.adaptors.gp2gp.ehr.exception.EhrMapperException;
+
+@Slf4j
+@AllArgsConstructor
+public class AgentDirectory {
+    private final RandomIdGeneratorService randomIdGeneratorService;
+    private final Map<AgentKey, String> agentDirectory = new HashMap<>();
+    private final Bundle inputBundle;
+
+    public Set<Map.Entry<AgentKey, String>> getEntries() {
+        return agentDirectory.entrySet();
+    }
+
+    public String getAgentId(Reference reference) {
+        if (reference.hasReferenceElement()) {
+            String resourceType = reference.getReferenceElement().getResourceType();
+            if (ResourceType.Practitioner.name().equals(resourceType)) {
+                return putIfAbsent(buildAgentKeyForPractitioner(reference));
+            } else if (ResourceType.Organization.name().equals(resourceType)) {
```

Comment:
Can an agent reference be to an organisation only? I though it must always be to practitioner. @dmcavenue1 ?

Reply:
I believe agentOrganization related to managingOgranization of Patient

Reply:
In public demonstrator, our test patient has Observations and DiagnosticReport that has single Organization reference in performer. Is it wrong?

Reply:
@mattd-kainos, Checked it with @dmcavenue1 and Observation and DiagnosticReport can have single Organization but it will be a corner case. And for now the mapping should behave as bellow:

- Organization referenced by Patient.managingOrganization is mapped once, and at first step as agentOrganization in agentDirectory
```
  <part typeCode="PART">
    <Agent classCode="AGNT">
        <id root="5E496953-065B-41F2-9577-BE8F2FBD0757" />
        <agentOrganization determinerCode="INSTANCE" classCode="ORG">
                        <id extension="A82038" root="2.16.840.1.113883.2.1.4.3"/>
            <name>TEMPLE SOWERBY MEDICAL PRACTICE</name>
            <telecom value="01133800000" use="WP"/>
            <addr use="WP">
                <streetAddressLine>Fulford Grange</streetAddressLine>
                <streetAddressLine>Micklefield Lane</streetAddressLine>
                <streetAddressLine>Rawdon</streetAddressLine>
                <streetAddressLine>Rawdon</streetAddressLine>
                <streetAddressLine>Leeds</streetAddressLine>
                <postalCode>LS19 6BA</postalCode>
            </addr>
        </agentOrganization>
    </Agent>
</part>
```
- then if any Observation or DiagnosticReport has reference to Organization only and that Organization is the same as in Patient.managingOrganization it will reference the one mapped to agentOrganization
- if any Observation or DiagnosticReport has reference to Organization only and that Organization is not in Patient.managingOrganization the agent is created with representedOrganization

```
<part typeCode="PART">
    <Agent classCode="AGNT">
        <id root="6D340A1B-BC15-4D4E-93CF-BBCB5B74DF73"/>
        <representedOrganization classCode="ORG" determinerCode="INSTANCE">
            <id extension="A82038" root="2.16.840.1.113883.2.1.4.3"/>
            <name>TEMPLE SOWERBY MEDICAL PRACTICE</name>
            <telecom value="01133800000" use="WP"/>
            <addr use="WP">
                <streetAddressLine>Fulford Grange</streetAddressLine>
                <streetAddressLine>Micklefield Lane</streetAddressLine>
                <streetAddressLine>Rawdon</streetAddressLine>
                <streetAddressLine>Rawdon</streetAddressLine>
                <streetAddressLine>Leeds</streetAddressLine>
                <postalCode>LS19 6BA</postalCode>
            </addr>
        </representedOrganization>
    </Agent>
</part>

```